# Core Linear Algebra Operations
**These operations form the computational backbone of machine learning models.**

**1. Addition / Subtraction**

Description: Element-wise addition or subtraction of matrices/vectors of the same shape.

Math:
C = A + B or C = A - B

Use in ML: Combine gradients, loss updates, feature scaling.

In [59]:
# Python Equivalent
import numpy as np
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])
c_add = A + B
c_sub = A - B
print("Matrix Addition : \n{}, \nMatrix Subtraction: \n{}".format(c_add,c_sub))

Matrix Addition : 
[[ 6  8]
 [10 12]], 
Matrix Subtraction: 
[[-4 -4]
 [-4 -4]]


**2. Scalar Multiplication**
Description: Multiply every element of a vector or matrix by a scalar value.

Math:

𝑐
𝐴
=
𝜆
⋅
𝐴

Use in ML: Learning rate scaling, normalization, weighting.

In [48]:
# Python Equivalent
A = np.array([[1, 2], [3, 4]])
C = 2 * A
print("Scalar Multiplication of A:",C)

Scalar Multiplication of A: [[2 4]
 [6 8]]


**3. Matrix Multiplication**

Description:  Multiply row of A with column of B

Formula: If A is (m×n), B is (n×p), then C = A × B is (m×p)

Example:

𝐴
=
[
1
2
3
4
]
,
𝐵
=
[
5
6
]
⇒
𝐶
=
𝐴
⋅
𝐵
=
[
(
1
×
5
+
2
×
6
)
(
3
×
5
+
4
×
6
)
]
=
[
17
39
]


Use in ML: Learning rate scaling, normalization, weighting.

In [49]:
# Python Equivalent
A = np.array([[1, 2], [3, 4]])
B = np.array([[2, 0], [1, 2]])
C = A @ B  # or np.dot(A, B)

print("Matrix Multiplication:/n", C)

Matrix Multiplication:/n [[ 4  4]
 [10  8]]


**4. Dot Product (Inner Product)**

Description: Multiply two vectors element-wise and sum the result.

Math:

𝑎
⃗
⋅
𝑏
⃗
=
∑
𝑎
𝑖
𝑏
𝑖

Example:

𝑎
=
[
1
,
2
]
,
𝑏
=
[
3
,
4
]
⇒
1
×
3
+
2
×
4
=
11

Use in ML: Similarity, projections, angle between vectors.


In [50]:
# Python Equivalent
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
dot = np.dot(a, b)

print("Dot product Inner:\n",dot)

Dot product Inner:
 32


**5. Outer Product**

Description: Multiply a column vector by a row vector to get a matrix. All combinations of aᵢ * bⱼ

Math:

𝑎
⃗
⊗
𝑏
⃗
=
𝑎
𝑏
𝑇

Example:

𝑎
=
[
1
,
2
]
,
𝑏
=
[
3
,
4
]
⇒
[
1
×
3
1
×
4
2
×
3
2
×
4
]
=
[
3
4
6
8
]


Use in ML: Covariance matrices, tensor construction, feature interaction.

In [53]:
# Python Equivalent
a = np.array([1, 2])
b = np.array([3, 4])
outer = np.outer(a, b)

print("Outer Product:\n", outer)

Outer Product:
 [[3 4]
 [6 8]]


**6. Hadamard Product**
Description: Element-wise multiplication (not matrix multiplication). Multiply element-by-element (must be same shape)
Math:

𝐶
=
𝐴
∘
𝐵
,
𝐶
𝑖
𝑗
=
𝐴
𝑖
𝑗
⋅
𝐵
𝑖
𝑗

Example:

𝐴
=
[
1
2
3
4
]
,
𝐵
=
[
5
6
7
8
]
⇒
𝐴
∘
𝐵
=
[
5
12
21
32
]


Use in ML: Attention models, dropout masks, gate operations in RNNs.

In [54]:
# Python Equivalent
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])
hadamard = A * B

print("Hadamard:\n", hadamard)

Hadamard:
 [[ 5 12]
 [21 32]]


**7. Transpose**

Description: Flips a matrix over its diagonal; rows become columns.  Flip rows and columns

Math:

𝐴
𝑇

Example:

𝐴
=
[
1
2
3
4
]
⇒
𝐴
𝑇
=
[
1
3
2
4
]


Use in ML: Change data orientation, gradient computation, symmetry checks.

In [55]:
# Python Equivalent
A = np.array([[1, 2], [3, 4]])
A_T = A.T

print("Transpose Matrix:\n", A_T)

Transpose Matrix:
 [[1 3]
 [2 4]]


**8. Inverse**

Description: Matrix inverse such that A⁻¹ · A = I.

Math:

𝐴
−
1
⋅
𝐴
=
𝐼

Formula (2x2):

𝐴
=
[
𝑎
𝑏
𝑐
𝑑
]
,
𝐴
−
1
=
1
𝑎
𝑑
−
𝑏
𝑐
[
𝑑
−
𝑏
−
𝑐
𝑎
]

Example:

𝐴
=
[
1
2
3
4
]
⇒
𝐴
−
1
=
[
−
2
1
1.5
−
0.5
]



Use in ML: Solving equations, normal equations in linear regression.

In [56]:
# Python Equivalent

A = np.array([[1, 2], [3, 4]])
A_inv = np.linalg.inv(A)
print("Inverse:\n", A_inv)

Inverse:
 [[-2.   1. ]
 [ 1.5 -0.5]]


**9. Determinant**
Description: Scalar measure of a matrix’s scaling factor or volume.

Math:

det
(
𝐴
)

Example:

𝐴
=
[
1
2
3
4
]
⇒
det
(
𝐴
)
=
1
×
4
−
2
×
3
=
−
2

Use in ML: Inversion validity, multivariate normal density, area/volume scaling.

In [57]:
# Python Equivalent
A = np.array([[1, 2], [3, 4]])
det = np.linalg.det(A)

print("Determinant", det)

Determinant -2.0000000000000004


**10. Norm**

Description: Length or magnitude of a vector.

Math:

∥
𝑥
⃗
∥
=
∑
𝑥
𝑖
2

(L2 Norm):

∥
𝑎
⃗
∥
=
∑
𝑎
𝑖
2

Example:

𝑎
=
[
3
,
4
]
⇒
3
2
+
4
2
=
5


Use in ML: Regularization (L1, L2), measuring distances.


In [58]:
# Python Equivalent
x = np.array([3, 4])
norm = np.linalg.norm(x)
print("Norm output:\n", norm)

Norm output:
 5.0


**11. Trace**

Description: Sum of the diagonal elements of a square matrix.

Math:

tr
(
𝐴
)
=
∑
𝐴
𝑖
𝑖


Example:

𝐴
=
[
1
2
3
4
]
⇒
tr
(
𝐴
)
=
1
+
4
=
5

Use in ML: In optimization, kernel tricks, matrix properties.

In [60]:
# Python Equivalent
A = np.array([[1, 2], [3, 4]])
trace = np.trace(A)

print("Trace",trace)

Trace 5


**12. Rank**

Description: Number of linearly independent rows or columns.

Math:
rank
(
𝐴
)

Example:

𝐴
=
[
1
2
2
4
]
⇒
Rank
(
𝐴
)
=
1

Use in ML: Data dimensionality, invertibility, linear dependence.


**13. Diagonalization**
Description: Concept: Convert a square matrix to diagonal using eigenvalues

**Formula:**

𝐴
=
𝑃
𝐷
𝑃
−
1
,
𝐷
=
diag
(
𝜆
1
,
𝜆
2
,
.
.
.
)

Where:
- P is the matrix of eigenvectors
- D is a diagonal matrix of eigenvalues

**Eigen Values:**

A=[[4,2],[1,3]]

det(A−λI)=0

Where
𝐼
I is the identity matrix:

det([
4−λ
2
​
  
1
3−λ
​
 ])=0

Calculate determinant:

(
4
−
𝜆
)
(
3
−
𝜆
)
−
(
2
)
(
1
)
=
0
⇒
𝜆
2
−
7
𝜆
+
10
=
0

Solve:

𝜆
=
5
,
𝜆
=
2

**So, the eigenvalues are 5 and 2.**


**Eigen Vectors:**

For each λ, solve:

(
𝐴
−
𝜆
𝐼
)
𝑣
⃗
=
0

For λ = 5:

𝐴
−
5
𝐼
=
[
−
1
1
2
−
2
]

Solving:

[
−
1
1
2
−
2
]
⋅
[
𝑥
𝑦
]
=
[
0
0
]

Gives x = y, so eigenvector is:

𝑣
1
=
[
1
1
]

For λ = 2:

𝐴
−
2
𝐼
=
[
2
1
2
1
]

Solving:

2
𝑥
+
𝑦
=
0
⇒
𝑦
=
−
2
𝑥

**So eigenvector:**

𝑣
2
=
[
1
−
2
]



Use in ML: PCA, spectral clustering, stability analysis.

In [61]:
# Python Equivalent
A = np.array([[4, 1], [1, 4]])
eigvals, eigvecs = np.linalg.eig(A)

print("Eigen Values : {},\nEigen Vectors: {}".format(eigvals, eigvecs))

Eigen Values : [5. 3.],
Eigen Vectors: [[ 0.70710678 -0.70710678]
 [ 0.70710678  0.70710678]]


# Summarize

| Operation               | Explanation & Example                                | NumPy Code                 | Output                         |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
| 1. Addition/Subtraction | Element-wise operation on same-shape matrices        |                            |                                |
|                        | A = [[1, 2], [3, 4]]                                  | A + B                      | [[ 5, 7],                      |
|                        | B = [[4, 5], [6, 7]]                                  |                            |  [ 9, 11]]                     |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
| 2. Scalar Multiplication| Multiply each element by scalar                      | 3 * A                      | [[ 3, 6],                      |
|                        | A = [[1, 2], [3, 4]]                                  |                            |  [ 9, 12]]                     |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
| 3. Matrix Multiplication| Dot product: (m×n) • (n×p) = (m×p)                   | A @ B or np.dot(A, B)      | [[19],                         |
|                        | A = [[1, 2, 3], [4, 5, 6]]                            |                            |  [46]]                         |
|                        | B = [[1], [2], [3]]                                   |                            |                                |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
| 4. Dot Product (Inner)  | Sum of element-wise product of vectors               | np.dot(a, b)               | 11                             |
|                        | a = [1, 2], b = [3, 4]                                |                            |                                |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
| 5. Outer Product        | Creates a matrix from vector outer product           | np.outer(a, b)             | [[ 3, 4],                      |
|                        | a = [1, 2], b = [3, 4]                                |                            |  [ 6, 8]]                      |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
| 6. Hadamard Product     | Element-wise multiplication                          | A * B                      | [[ 4, 10],                     |
|                        | A = [[1, 2], [3, 4]]                                  |                            |  [18, 28]]                     |
|                        | B = [[4, 5], [6, 7]]                                  |                            |                                |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
| 7. Transpose            | Flip matrix over diagonal                            | A.T                        | [[1, 3],                       |
|                        | A = [[1, 2], [3, 4]]                                  |                            |  [2, 4]]                       |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
| 8. Inverse              | A⁻¹ such that A⁻¹ @ A = I                            | np.linalg.inv(A)           | [[-2. ,  1. ],                 |
|                        | A = [[1, 2], [3, 4]]                                  |                            |  [ 1.5, -0.5]]                 |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
| 9. Determinant          | Scalar measure of matrix area/volume                 | np.linalg.det(A)           | -2.0000000000000004           |
|                        | A = [[1, 2], [3, 4]]                                  |                            |                                |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
|10. Norm (Length)        | Length of a vector                                   | np.linalg.norm(a)          | 5.0                            |
|                        | a = [3, 4]                                            |                            |                                |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
|11. Trace                | Sum of diagonal elements                             | np.trace(A)                | 5                              |
|                        | A = [[1, 2], [3, 4]]                                  |                            |                                |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
|12. Rank                 | No. of linearly independent rows/columns             | np.linalg.matrix_rank(A)   | 2                              |
|                        | A = [[1, 2], [2, 4]]                                  |                            |                                |
|------------------------|-------------------------------------------------------|----------------------------|--------------------------------|
|13. Diagonalization      | A = P D P⁻¹, D is diagonal with eigenvalues          | np.linalg.eig(A)           | eigenvalues: [5.372, -0.372]   |
|                        | A = [[1, 2], [2, 3]]                                  |                            | eigenvectors: [[...], [...]]  |
